## Processing of metagenomics data
In this workshop, we follow the steps of the Ning et. al. manuscript ["Microbiome and metabolome features in inflammatory bowel disease via multi-omics integration analyses across cohorts"](https://www.nature.com/articles/s41467-023-42788-0) of processing metagenomics sequences using: 
1. Fastqc
2. Trimmomatic
3. Bowtie2
4. Metaphlan
5. And combining steps 1-3 into one pipeline using KneadData

Example FastQC files are:
Raw files:
1. [SRR6468499_1](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/raw/SRR6468499_1_fastqc.html); [SRR6468499_2](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/raw/SRR6468499_2_fastqc.html)
2. [SRR6468511_1](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/raw/SRR6468511_1_fastqc.html) [SRR6468511_2](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/raw/SRR6468511_2_fastqc.html)

Trimmed fils:
1. [SRR6468499_1](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/trimmed/SRR6468499_1.paired_fastqc.html); [SRR6468499_2](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/trimmed/SRR6468499_2.paired_fastqc.html)
2. [SRR6468511_1](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/trimmed/SRR6468511_1.paired_fastqc.html) [SRR6468511_2](https://pollytikhonova.github.io/DAWG_workshop/2024-2025/W2_metagenomics_processing/fastqc_reports/trimmed/SRR6468511_2.paired_fastqc.html)

